woocommerce/woocommerce-ios
# [Mobile Payments] Prepare for providing feedback of payment collection process on the UI

Closes #3842 , #3843 

⚠️ This PR is against the feature branch implementing support for Card Present Payments, not against develop ⚠️
⚠️ Please note there are several parts of the implementation that are not production ready.⚠️

We continue building our prototype (for reference, p91TBi-4q4-p2), this time, by implementing enough code to propagate certain events dispatched by the card reader to the UI. In particular:

* When the card reader requests a card to be inserted / tapped / swiped.
* When the card reader requests the card to be removed.
* When the card reader requests any string to be presented on screen.

These events are generated by the Stripe Terminal SDK. For some of those events, the Stripe Terminal SDK also provides a user facing string (see [this](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(cm)stringFromReaderInputOptions:) and [this](https://stripe.dev/stripe-terminal-ios/docs/Classes/SCPTerminal.html#/c:objc(cs)SCPTerminal(cm)stringFromReaderDisplayMessage:)). We are passing those strings to the UI as well, although, in my opinion, and from what I've seen in my tests, the 
value they provide is debatable. 

## Changes
* Add a couple of cases to CardReaderEventType
* Add a new model object CardReaderEvent that wraps a CardReaderEventType and a user facing string
* Re-implement the methods declared in the conformance of StripeCardReaderService to ReaderDisplayDelegate so that they send new events via the reader events subject
* Update CardPresentPaymentStore to observe those events and update the API of the collectPayment method to propagate the events to the UI layer. See #3926 for some known issues in the implementation of th CardPresentPaymentStore.
* Add just enough code to OrderDetailsViewModel and OrderDetailsViewController to present some feedback on the UI, so that testing the PR is a bit easier. This code will definitely be removed later on, and the APIs are not final. This is part of the PR only so that there is visible feedback on the UI of the changes.

## How to test
Testing is a bit complicated because:
1. Requires setting up a store for [Card Present Payment Testing](https://github.com/woocommerce/woocommerce-ios/blob/issue/3748-intent-service/docs/stripe-tests.md#integration-tests) (also P91TBi-4BH-p2 for more specific instructions)  
2. Requires an external hardware reader.

Once the store is setup and there is a hardware reader available:
1. Checkout the branch. Run `bundle exec pod install`. 
2. Build and run on a device. 
3. Switch an external card reader on. The only card reader supported by now is the BBPOS Chipper 2X BT.
4. Log in to an account that matches a store that has been configured for testing.
5. Navigate to Settings in the app. (Tap the gear button in the top right)
6. In the "Store Settings" section, tap "Manage Card Reader".
7. Make sure the card reader is on and the blue light is blinking (we don't do any error handling yet, so it's important that things are setup for the happy path)
8. Tap "Connect card reader"
9. Wait until the app suggests one reader to connect to.
10. Tap "Connect to reader"
11. Make sure the reader light turns solid blue.
12. Navigate to an order (any order with payment method Cash On Delivery).
13. Scroll down to the "Collect Payment" button
14. Tap the button. Follow the instructions on screen


This is a video of the "before" (notice there is only one alert at the end of the process):
https://user-images.githubusercontent.com/2722505/111704327-54016380-8815-11eb-9feb-05c78f3a9eae.mov

And this is how it should look now (notice there is an alert prompting to tap the card, and that said alert is dismissed after tapping the card):

https://user-images.githubusercontent.com/2722505/113864644-24d97480-9779-11eb-93a8-df4e33c25c93.mov


Update release notes:
